pydantic_evals/pydantic_evals/datasets.py

```diff
@@ -0,0 +1,376 @@
+from __future__ import annotations as _annotations
+
+import functools
+import inspect
+import sys
+from pathlib import Path
+from typing import Any, ClassVar, Generic, Literal, NotRequired, Self, Union
+
+from pydantic._internal import _typing_extra
+
+if sys.version_info < (3, 11):
+    from exceptiongroup import ExceptionGroup
+else:
+    ExceptionGroup = ExceptionGroup
+
+import yaml
+from pydantic import BaseModel, Field, ValidationError
+from pydantic_core import to_json, to_jsonable_python
+from typing_extensions import TypedDict, TypeVar
+
+from .assessments.spec import Assessment, AssessmentFunction, AssessmentSpec, get_default_registry
+
+InputsT = TypeVar('InputsT', default=dict[str, Any])
+OutputT = TypeVar('OutputT', default=dict[str, Any])
+MetadataT = TypeVar('MetadataT', default=dict[str, Any])
+
+DEFAULT_DATASET_PATH = './test_cases.yaml'
+
+
+class DatasetRow(BaseModel, Generic[InputsT, OutputT, MetadataT], extra='forbid'):
+    """A single row of a "dataset", consisting of input, expected output, and metadata."""
+
+    name: str
+    inputs: InputsT
+    metadata: MetadataT
+    expected_output: OutputT | None = None
+    assessments: list[AssessmentSpec] = Field(default_factory=list)
+
+
+class EvaluationRow(BaseModel, Generic[InputsT, OutputT, MetadataT], extra='forbid'):
+    """A single row for evaluation."""
+
+    name: str
+    inputs: InputsT
+    metadata: MetadataT
+    expected_output: OutputT | None
+    assessments: list[Assessment[InputsT, OutputT, MetadataT]]
+
+
+F = TypeVar('F', bound=AssessmentFunction[Any, Any, Any])
+
+
+class Dataset(BaseModel, Generic[InputsT, OutputT, MetadataT], extra='forbid'):
+    """A dataset of test cases, each consisting of input, expected output, and metadata."""
+
+    rows: list[DatasetRow[InputsT, OutputT, MetadataT]]
+    default_assessments: list[AssessmentSpec] = Field(default_factory=list)
```

Comment:
'default' sounds like they can be overridden, are we going to allow not running the default assessments for some cases?

Reply:
agreed I think it's a terrible name, I changed it to scorers to match other libraries I've seen but I don't think that name's right either ...